mirrord/layer/src/socket.rs

```diff
@@ -499,6 +537,40 @@ impl OutgoingSelector {
             .collect()),
         }
     }
+
+    /// Helper function that looks into the [`DNS_CACHE`] for `address`, so we can retrieve the
+    /// hostname and resolve it locally (when applicable).
+    ///
+    /// We only get here when the [`OutgoingSelector::Remote`] matched nothing, thus
+    /// `selected_address` comes in as `None`, or when the [`OutgoingSelector::Local`] matched on
+    /// something, so we have `Some(selected_address)`.
+    ///
+    /// Returns 1 of 3 possibilities:
+    ///
+    /// 1. `address` is in [`DNS_CACHE`]: resolves the hostname locally, then return it as
+    /// [`ConnectionThrough::Local`];
+    /// 2. `address` is **NOT** in [`DNS_CACHE`]: check `REMOTE` and return the valid local address;
+    #[tracing::instrument(level = "trace", ret)]
+    fn get_address_to_connect<const REMOTE: bool>(
+        address: SocketAddr,
+        selected_address: Option<SocketAddr>,
+    ) -> Detour<ConnectionThrough> {
+        if let Some((cached_hostname, port)) = DNS_CACHE
+            .get(&SocketAddr::from((address.ip(), 0)))
+            .map(|addr| (addr.value().clone(), address.port()))
+        {
+            let _guard = DetourGuard::new();
+            let locally_resolved = format!("{cached_hostname}:{port}")
+                .to_socket_addrs()?
+                .find(SocketAddr::is_ipv4)?;
+
+            Detour::Success(ConnectionThrough::Local(locally_resolved))
+        } else if REMOTE {
+            Detour::Success(ConnectionThrough::Local(address))
+        } else {
+            Detour::Success(ConnectionThrough::Local(selected_address.unwrap()))
```

Comment:
This unwrap, while ok in this version of the code, would not be needed if we made two non-generic functions for the two different cases, which speaks for doing that, in my opinion.

Reply:
Actually, think we don't need this stuff, `selected_address` should just be `address`, and there is no need to check `REMOTE`.

Trying it out.

Reply:
Yup, simplified the function.